Use UserContentURLPattern as the backing for WebExtensionMatchPattern.

xeenon opened the pull request:
#### 70c03f5cf50d47716f2ef5f2b11c6452f0a9f216
<pre>
Use UserContentURLPattern as the backing for WebExtensionMatchPattern.
<a href="https://bugs.webkit.org/show_bug.cgi?id=250624">https://bugs.webkit.org/show_bug.cgi?id=250624</a>

Reviewed by Brian Weinstein.

WebExtensionMatchPattern has been made a thin wrapper around UserContentURLPattern by creating an internal pattern and using
UserContentURLPattern&apos;s matching methods to support WebExtensionMatchPattern&apos;s matching options (ignoring scheme, path, etc).

UserContentURLPattern now only matches HTTP-family schemes with the &apos;*&apos; scheme for Web Extensions compatibility. This change
impacted a few API tests that expected &apos;*://*/*&apos; to match custom schemes. I&apos;ve changed those tests to use the expected scheme.
Extensions and apps are not be impacted since HTTP-family URLs match like before, and only Safari specific SPIs use or expose
the UserContentURLPattern class and matching.

UserContentURLPattern now normalizes &apos;file://localhost/*&apos; to &apos;file:///*&apos; to match how URLParser parses URLs, so that matching
these URLs work universally. Additionally, hosts are now supported for file URLs, as expected by some unit tests.

Creating a pattern with the three components (scheme, host, path) is now supported.

This is significantly more efficient for WebExtensionMatchPattern. It has been observed to be 17x faster in a micro-benchmark.

* Source/WebCore/page/UserContentURLPattern.cpp:
(WebCore::UserContentURLPattern::UserContentURLPattern): Added.
(WebCore::UserContentURLPattern::operator== const): Added.
(WebCore::UserContentURLPattern::normalizeHostAndSetMatchSubdomains): Added.
(WebCore::UserContentURLPattern::parse): Use normalizeHostAndSetMatchSubdomains and return errors instead of bool.
(WebCore::UserContentURLPattern::originalHost const): Added.
(WebCore::UserContentURLPattern::matches const):
(WebCore::UserContentURLPattern::matchesScheme const):
(WebCore::UserContentURLPattern::matchesHost const):
(WebCore::MatchTester::test):
(WebCore::UserContentURLPattern::matchesPath const):
* Source/WebCore/page/UserContentURLPattern.h:
(WebCore::UserContentURLPattern::UserContentURLPattern):
(WebCore::UserContentURLPattern::isValid const):
(WebCore::UserContentURLPattern::error const):
(WebCore::UserContentURLPattern::matchesHost const):
(WebCore::UserContentURLPattern::matchesPath const):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMatchPattern.mm:
(-[_WKWebExtensionMatchPattern scheme]): Return nil if the String is null.
(-[_WKWebExtensionMatchPattern host]): Ditto.
(-[_WKWebExtensionMatchPattern path]): Ditto.
(-[_WKWebExtensionMatchPattern matchesURL:options:]): Add assert for option that can&apos;t be used.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::registerCustomURLScheme):
(WebKit::WebExtensionMatchPattern::getOrCreate):
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern):
(WebKit::WebExtensionMatchPattern::isSupported const):
(WebKit::WebExtensionMatchPattern::operator== const):
(WebKit::WebExtensionMatchPattern::scheme const): Added.
(WebKit::WebExtensionMatchPattern::host const): Added.
(WebKit::WebExtensionMatchPattern::path const): Added.
(WebKit::WebExtensionMatchPattern::stringWithScheme const):
(WebKit::WebExtensionMatchPattern::expandedStrings const):
(WebKit::WebExtensionMatchPattern::matchesAllHosts const):
(WebKit::WebExtensionMatchPattern::isValidScheme):
(WebKit::WebExtensionMatchPattern::matchesURL const):
(WebKit::WebExtensionMatchPattern::matchesPattern const):
(WebKit::WebExtensionMatchPattern::schemeMatches const):
(WebKit::WebExtensionMatchPattern::hostMatches const):
(WebKit::WebExtensionMatchPattern::pathMatches const):
(WebKit::WebExtensionMatchPattern::isValidHost): Deleted.
(WebKit::WebExtensionMatchPattern::isValidPath): Deleted.
(WebKit::WebExtensionMatchPattern::parse): Deleted.
(WebKit::WebExtensionMatchPattern::matchesURL): Deleted.
(WebKit::WebExtensionMatchPattern::matchesPattern): Deleted.
(WebKit::WebExtensionMatchPattern::schemeMatches): Deleted.
(WebKit::WebExtensionMatchPattern::hostMatches): Deleted.
(WebKit::WebExtensionMatchPattern::pathMatches): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionMatchPattern.mm:
(TestWebKitAPI::TEST): Change some expectations to match new results for file URLs and URL percent encodings.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentExtensionStore.mm:
(navigationDelegateAllowingActiveActionsOnTestHost): Changed wildcard scheme in pattern to &apos;testscheme&apos; scheme expected by the test.
(TestWebKitAPI::TEST_F(WKContentRuleListStoreTest, NullPatternSet)): Ditto.
(TestWebKitAPI::TEST_F(WKContentRuleListStoreTest, ExtensionPath)): Ditto.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKURLSchemeHandler-1.mm:
(TestWebKitAPI::TEST(URLSchemeHandler, DisableCORSCanvas)): Changed wildcard scheme in pattern to &apos;cors&apos; scheme expected by the test.

Canonical link: <a href="https://commits.webkit.org/259009@main">https://commits.webkit.org/259009@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/048e02d92be3ec8d627101d88e62446164949b2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103649 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12765 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/36598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112879 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107601 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3662 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/95891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109420 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/36598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/95891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/92447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/36598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/95891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6134 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/36598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/6311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/12293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/36598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8068 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3291 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebCore/page/UserContentURLPattern.cpp:
```diff
@@ -61,71 +98,92 @@ bool UserContentURLPattern::matchesPatterns(const URL& url, const Vector<String>
-bool UserContentURLPattern::parse(StringView pattern)
+void UserContentURLPattern::normalizeHostAndSetMatchSubdomains()
+{
+    ASSERT(!m_matchSubdomains);
```
It might be nice if `m_matchSubdomains` had some third state that indicated it hadn't been configured yet, then this assert would be a little more actionable.

Really, you're just trying to assert this method doesn't get called more than once, correct?

xeenon replied:
Correct, this ASSERT is just to avoid setting `m_matchSubdomains` twice in the method, since it will start as false for a fresh object.

b-weinstein commented on Source/WebCore/page/UserContentURLPattern.h:
```diff
@@ -34,34 +35,65 @@ class UserContentURLPattern {
     String m_path;
 
-    bool m_invalid { true };
+    Error m_error { Error::None };
```
Previously an uninitialized UserConetntURLPattern would be invalid - now it would be considered valid (since there is no error). Is this safe? Or should we start it in an invalid state so no one tries to use an uninitialized one of these?

xeenon replied:
Good catch! Maybe I'll add another error state to the enum for that case.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMatchPattern.mm:
```diff
@@ -175,16 +175,22 @@ - (NSString *)debugDescription
 {
+    if (_webExtensionMatchPattern->path().isNull())
+        return nil;
     return _webExtensionMatchPattern->path();
```
It seems like these three methods could use a local to get their relevant variables, but definitely not necessary.

xeenon committed:
Use UserContentURLPattern as the backing for WebExtensionMatchPattern.
https://bugs.webkit.org/show_bug.cgi?id=250624

Reviewed by Brian Weinstein.

WebExtensionMatchPattern has been made a thin wrapper around UserContentURLPattern by creating an internal pattern and using
UserContentURLPattern's matching methods to support WebExtensionMatchPattern's matching options (ignoring scheme, path, etc).

UserContentURLPattern now only matches HTTP-family schemes with the '*' scheme for Web Extensions compatibility. This change
impacted a few API tests that expected '*://*/*' to match custom schemes. I've changed those tests to use the expected scheme.
Extensions and apps are not be impacted since HTTP-family URLs match like before, and only Safari specific SPIs use or expose
the UserContentURLPattern class and matching.

UserContentURLPattern now normalizes 'file://localhost/*' to 'file:///*' to match how URLParser parses URLs, so that matching
these URLs work universally. Additionally, hosts are now supported for file URLs, as expected by some unit tests.

Creating a pattern with the three components (scheme, host, path) is now supported.

This is significantly more efficient for WebExtensionMatchPattern. It has been observed to be 17x faster in a micro-benchmark.

* Source/WebCore/page/UserContentURLPattern.cpp:
(WebCore::UserContentURLPattern::UserContentURLPattern): Added.
(WebCore::UserContentURLPattern::operator== const): Added.
(WebCore::UserContentURLPattern::normalizeHostAndSetMatchSubdomains): Added.
(WebCore::UserContentURLPattern::parse): Use normalizeHostAndSetMatchSubdomains and return errors instead of bool.
(WebCore::UserContentURLPattern::originalHost const): Added.
(WebCore::UserContentURLPattern::matches const):
(WebCore::UserContentURLPattern::matchesScheme const):
(WebCore::UserContentURLPattern::matchesHost const):
(WebCore::MatchTester::test):
(WebCore::UserContentURLPattern::matchesPath const):
* Source/WebCore/page/UserContentURLPattern.h:
(WebCore::UserContentURLPattern::UserContentURLPattern):
(WebCore::UserContentURLPattern::isValid const):
(WebCore::UserContentURLPattern::error const):
(WebCore::UserContentURLPattern::matchesHost const):
(WebCore::UserContentURLPattern::matchesPath const):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMatchPattern.mm:
(-[_WKWebExtensionMatchPattern scheme]): Return nil if the String is null.
(-[_WKWebExtensionMatchPattern host]): Ditto.
(-[_WKWebExtensionMatchPattern path]): Ditto.
(-[_WKWebExtensionMatchPattern matchesURL:options:]): Add assert for option that can't be used.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::registerCustomURLScheme):
(WebKit::WebExtensionMatchPattern::getOrCreate):
(WebKit::WebExtensionMatchPattern::WebExtensionMatchPattern):
(WebKit::WebExtensionMatchPattern::isSupported const):
(WebKit::WebExtensionMatchPattern::operator== const):
(WebKit::WebExtensionMatchPattern::scheme const): Added.
(WebKit::WebExtensionMatchPattern::host const): Added.
(WebKit::WebExtensionMatchPattern::path const): Added.
(WebKit::WebExtensionMatchPattern::stringWithScheme const):
(WebKit::WebExtensionMatchPattern::expandedStrings const):
(WebKit::WebExtensionMatchPattern::matchesAllHosts const):
(WebKit::WebExtensionMatchPattern::isValidScheme):
(WebKit::WebExtensionMatchPattern::matchesURL const):
(WebKit::WebExtensionMatchPattern::matchesPattern const):
(WebKit::WebExtensionMatchPattern::schemeMatches const):
(WebKit::WebExtensionMatchPattern::hostMatches const):
(WebKit::WebExtensionMatchPattern::pathMatches const):
(WebKit::WebExtensionMatchPattern::isValidHost): Deleted.
(WebKit::WebExtensionMatchPattern::isValidPath): Deleted.
(WebKit::WebExtensionMatchPattern::parse): Deleted.
(WebKit::WebExtensionMatchPattern::matchesURL): Deleted.
(WebKit::WebExtensionMatchPattern::matchesPattern): Deleted.
(WebKit::WebExtensionMatchPattern::schemeMatches): Deleted.
(WebKit::WebExtensionMatchPattern::hostMatches): Deleted.
(WebKit::WebExtensionMatchPattern::pathMatches): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionMatchPattern.mm:
(TestWebKitAPI::TEST): Change some expectations to match new results for file URLs and URL percent encodings.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentExtensionStore.mm:
(navigationDelegateAllowingActiveActionsOnTestHost): Changed wildcard scheme in pattern to 'testscheme' scheme expected by the test.
(TestWebKitAPI::TEST_F(WKContentRuleListStoreTest, NullPatternSet)): Ditto.
(TestWebKitAPI::TEST_F(WKContentRuleListStoreTest, ExtensionPath)): Ditto.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKURLSchemeHandler-1.mm:
(TestWebKitAPI::TEST(URLSchemeHandler, DisableCORSCanvas)): Changed wildcard scheme in pattern to 'cors' scheme expected by the test.

Canonical link: https://commits.webkit.org/259009@main
